test: use installed catalogs when test-catalog is not located at build dir

yuwata opened the pull request:
Fixes #10045.

mbiebl commented:
Thank @yuwata. I see the installed test passed now (in the root-unittests autopkgtest) whereas it previously failed:
```
PASS: test-catalog
```
vs
```
====== test-catalog =======
[/tmp/test-catalog.iXkvgu:1] Got payload before ID.
[/tmp/test-catalog.i1u7ds:1] language too long.
No items in catalog.
No items in catalog.
No items in catalog.
Assertion 'r >= 0' failed at ../src/journal/test-catalog.c:226, function main(). Aborting.
Aborted
EXFAIL: test-catalog
```
So this particular issue of #10045 has been fixed.

Do you have an idea how we can get rid of the absolute paths in the binaries? I notice that this PR still encodes an absolute build path in the binary via ```ABS_BUILD_DIR```? Then again, we can/should probably split this into a separate issue. Looks like ```ABS_BUILD_DIR``` is used in other places as well, like shared/tests.c.

mbiebl commented on src/test/meson.build:
```diff
@@ -757,7 +757,7 @@ tests += [
           libxz,
           liblz4],
-         '', '', '-DCATALOG_DIR="@0@"'.format(build_catalog_dir)],
+         '', '', '-DCATALOG_DIR="@0@"'.format(catalogdir)],
```
I think we can drop
```
build_catalog_dir = meson.current_build_dir()
```
from catalog/meson.build as build_catalog_dir is no longer used anywhere in the code.

fsateler replied:
Wouldn't this change make the test use the installed catalogs instead of the source ones?

mbiebl replied:
@fsateler CATALOG_DIR is only used when executed from the build directory.

> Wouldn't this change make the test use the installed catalogs instead of the source ones?

not afaics, the patch changes the logic by trying to detect if the binary was exectuted from the build directory and if so constructs the catalog dir via ```catalog_dir = path_join(NULL, ABS_BUILD_DIR, "catalog")```

filbranden commented on src/journal/test-catalog.c:
```diff
@@ -202,14 +201,27 @@ static void test_catalog_file_lang(void) {
+        /* If test-catalog is located at the build directory, then use catalogs in that.
+         * If it is not, e.g. installed by systemd-tests package, then use installed catalogs. */
+        assert_se(readlink_and_make_absolute("/proc/self/exe", &exedir) >= 0);
+        if (path_startswith(exedir, ABS_BUILD_DIR)) {
```
You might want to refactor this code into a shared function such as `bool is_test_running_from_builddir()`, which can then be used in the two places that want to detect that:

https://github.com/systemd/systemd/blob/da0da5eccf5f20874111e0681b111704bf9a1c92/src/shared/tests.c#L38-L41

I'm not super thrilled about using `ABS_BUILD_DIR` to detect this, but given this is pre-existing, it's probably fine to keep detecting it this way. I'm wondering if there's a better way to detect this, such as an environment that gets exported by `ninja` etc., but it's possible this would be called stand-alone on the test environment too, so maybe this is as good as we can do it...

Cheers,
Filipe

mbiebl replied:
@filbranden I would very much like to see this. Passing the build directory via an env var when running "ninja test" and not encoding this in the binary itself, would help with making the reproducible. See also my comments in
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=908365#42

One downside of this would be, that you'd have to remember to set that env var when you execute the test-binary manually.

mbiebl replied:
The Debian reproducibility effort operates on binary packages. If we drop systemd-tests, there would be not need (from the Debian POV) to get rid of the use of ```ABS_BUILD_DIR```.

IIRC you can declare an autopkgtest as both ```needs_root``` and ```needs_build```. The later would result in the package being built as part of the autopkgtest run. So theoretically this should be possible. Keep in mind though, that a full build requires a lot of ressurces (especially the Debian build which builds both a udeb and non-udeb variant).
Also keep in mind, that whenever a (reverse)dependency of systemd changes, the Debian CI (and I think the Ubuntu CI as well) will trigger a full autopkgtest run [1].
If that would include a full rebuild each time, this would suddenly get a lot more expensive.
Installed tests are useful if you want to execute them often.

[1]  https://ci.debian.net/packages/s/systemd/testing/amd64/

mbiebl replied:
being able to run the tests as root is one aspect, being able to run them often (as explained above) is another

mbiebl replied:
Maybe https://wiki.gnome.org/Initiatives/GnomeGoals/InstalledTests does a better job at explainging it

evverx replied:
I'm not sure I understand what was explained above. Who needs to run the unit tests often and why? Nobody had needed that before f853c6efb509589c8e3a82caa3af2 was merged. Probably I missed what happened afterwards.

evverx replied:
I'm sorry. I didn't see the link you posted when I was writing the previous comment. I'll try to find the answer there. Or, more likely, I'll find what is written there unconvincing :-)

evverx replied:
Let me get this right. We complicated the tests because for some reason it wasn't possible to just run:
```sh
meson
ninja ... test
sudo ninja ... test
```
on Ubuntu CI, which, in turn, apparently broke the reproducibility of the Debian package. And all we have now is the ability to run the tests often according to the proposal that hasn't been approved yet and, judging by the absence of the files with the `[Test]` section in https://salsa.debian.org/systemd-team/systemd, this ability isn't quite useful at the moment.

mbiebl replied:
Those .test files are only necessary if you want to run the tests using GLibs test runner called gtester. We don't use those. The link to the GNOME proposal was just to show the motivation for installed tests

mbiebl replied:
anyway, this is really getting OT for this PR

evverx replied:
I wouldn't say that discussing the underlying issue is off-topic, but, if everybody is OK to complicate things by adding the unnecessary complexity, which is, in my opinion, very distro-specific, then, well, it's fine. It should be added somewhere.

mbiebl replied:
There is nothing distro-specific about using installed tests. But I feel like we are going in circles, so I'll stop at this point.

evverx replied:
I agree that there's nothing distro-specific in installed tests per se. However, to be really useful outside of the environments where there are special scripts like https://salsa.debian.org/systemd-team/systemd/blob/master/debian/tests/root-unittests, the tests should come with some kind of a runner which would run them. Without it they aren't that useful, which makes me think that they aren't widely used. Though, I may be wrong of course. Anyway, I'll stop stop here too. Thank you for the links. I've learnt something new today.

yuwata commented:
Updated.
- added the helper function suggested by @filbranden.
- dropped the unused variable `build_catalog_dir` suggested by @mbiebl.

I have not dropped the hard coded `ABS_BUILD_DIR` or `ABS_SRC_DIR` from tests. I think it should be done in a separate PR.

evverx commented:
I've been thinking about removing `ABS_BUILD_DIR` and the only solution I've come up with (apart from getting rid of the relocatable tests :-)) is to somehow mark the installed binaries so that it would be easy to distinguish them. Would it makes sense to add a suffix using something like
```
meson.add_install_script('sh', '-c', 'for n in $DESTDIR@0@/test-*; do mv $n ${n}_INSTALLED; done'.format(testsdir))
```
? That way we could tell whether a binary is outside of the build tree by looking at the end of the binary name, that is `test_is_running_from_builddir` would effectively become `!endswith("_INSTALLED")`. I can't see what could possibly go wrong here :-)

evverx commented:
Turns out removing `ABS_SRC_DIR` is feasible but much more complicated than I thought.

Are there reproducible packages out there where the tests are also relocatable? It would probably useful to take a look at them to see how it's done. (I hope I'm missing something obvious here and there is an easy way to do that).

mbiebl commented:
@evverx My idea was the following:
- Let the build system/meson set an env var path  when executing the tests in the build directory. This path would point at the src directory.
- get_testdata_dir() would be much simpler: It checks if the env var is set and is so, returns that path. Otherwise it would return the install path, i.e. ```/usr/lib/systemd/tests/testdata```

yuwata commented:
> - get_testdata_dir() would be much simpler: It checks if the env var is set and if so, returns that path. Otherwise it would return the install path, i.e. ```/usr/lib/systemd/tests/testdata```

I often run e.g., `test-execute` manually. I do not want to set `$SYSTEMD_TEST_DATA` every time when I run tests manually...

mbiebl commented:
> I often run e.g., test-execute manually. I do not want to set $SYSTEMD_TEST_DATA every time when I run tests manually...

It would then use the test-data from the installed system, which would probably work just fine in most cases. But yeah, I see the problem.

mbiebl commented:
@yuwata So here is a slightly different proposal:
- We let meson generate a test.config file in the build directory, which contains srcdir path
- ```test_is_running_from_builddir()``` is changed. We determine the exe path as before, but instead of comparing if it starts with ```ABS_BUILD_DIR```, we check if the directory contains a test.config file. This should work, as all test executables are generated in the same toplevel directory as the test.conf file.
- We can then read the src dir from test.config and pass that back to the caller of ```test_is_running_from_builddir()```.

mbiebl commented:
The failing bionic-am64 check is a timeout and looks unrelated. Merging the PR.
